Java/core/src/main/java/com/amazon/randomcutforest/genericstore/Store.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.genericstore;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.lang.reflect.Array;
+import java.util.BitSet;
+
+/**
+ * This class provides a fixed amount of storage slots for a given type. When an
+ * object is stored, the class returns an index value that can be used to
+ * retrieve the object later. The purpose of this class is to reduce the overall
+ * memory footprint of an applicaiton by allowing classes to keep index values
+ * instead of pointers. The index values used in this class are ints, which
+ * occupy 4 bytes (Java pointers occupy 8 bytes). For further compression, see
+ * the {@link SmallStore} class which uses 2 byte short integers as index
+ * values.
+ *
+ * @param <T> The type of value being stored.
+ */
+public class Store<T> {
+
+    private final int capacity;
+    private final T[] store;
+    private final int[] freeBlockStack;
+    private int freeBlockPointer;
+    private final BitSet occupied;
+
+    /**
+     * Create a new store with the given capacity.
+     * 
+     * @param capacity The maximum number of objects that can be added to this
+     *                 store.
+     */
+    public Store(Class<T> clazz, int capacity) {
+        checkArgument(capacity > 0, "capacity must be greater than 0");
+        this.capacity = capacity;
+
+        @SuppressWarnings("unchecked")
+        final T[] temp = (T[]) Array.newInstance(clazz, capacity);
```

Comment:
question. is temp variable required?

Reply:
This class will be removed in the next revision.